base/abstractarray.jl

```diff
@@ -230,18 +236,19 @@ end
 
 function copy!(dest::AbstractArray, doffs::Integer, src, soffs::Integer)
     if (doffs < 1) | (soffs < 1)
-        throw(BoundsError())
+        doffs < 1 && throw(BoundsError(dest, doffs))
+        throw(BoundsError(src, soffs))
     end
     st = start(src)
     for j = 1:(soffs-1)
-        done(src, st) && throw(BoundsError())
+        done(src, st) && throw(BoundsError(src, j))
```

Comment:
I'm not entirely sure about this.  In this method, `src` is a general iterable and *not* an abstract array (there are later specializations for abstract array sources).  Are there other places where we use bounds errors for iterables (that aren't arrays) terminating earlier than expected? If not, a more accurate message might be something along the lines of: `throw(ArgumentError("source has fewer elements than required, expected at least $soffs, got $j"))`.